README.rst
.. image:: https://github.com/xcir/libvmod-awsrest/actions/workflows/test.yml/badge.svg?branch=master
    :target: https://github.com/xcir/libvmod-awsrest/actions/workflows/test.yml

===================
vmod_awsrest
===================

-------------------------------
Varnish AWS REST API module
-------------------------------

:Author: Shohei Tanaka(@xcir)
:Date: 2021/10/19
:Version: 70.12
:Support Varnish Version: 5.0.x ~ 7.0.x
:Manual section: 3

SYNOPSIS
========

import awsrest;

Versioning(Source)
====================
[varnish-version].[library-version]

65.1 is v1 for Varnish6.5.x

Versioning(Package)
====================
[VRT-version].[Source-version]

120.65.1 is 65.1 for VRT12.0

============ ===============
VRT Version  Varnish Version 
------------ ---------------
14.0         7.0.x
13.0         6.6.x
12.0         6.5.x
11.0         6.4.x
10.0         6.3.x
9.0          6.2.x
8.0          6.1.x
7.1          6.0.4~6.0.x
7.0          6.0.0~6.0.3
============ ===============

For Varnish3.0.x
=================

See this link.
https://github.com/xcir/libvmod-awsrest/tree/varnish30

For Varnish4.0.x
=================

See this link.
https://github.com/xcir/libvmod-awsrest/tree/varnish40

For Varnish4.1.x
=================

See this link.
https://github.com/xcir/libvmod-awsrest/tree/varnish41

DESCRIPTION
===========

FUNCTIONS
============

v4_generic
------------------

Prototype
        ::

                v4_generic(
                    STRING service,               // [s3]
                    STRING region,                // [ap-northeast-1]
                    STRING access_key,            // [your access key]
                    STRING secret_key,            // [your secret key]
                    STRING session_token,         // [your session token (optional)]
                    STRING signed_headers,        // [host;]                                   x-amz-content-sha256;x-amz-date is appended by default.
                    STRING canonical_headers,     // [host:s3-ap-northeast-1.amazonaws.com\n]
                    BOOL   feature                // [false]                                   reserved param
                    )
Return value
	VOID
Description
	generate Authorization/x-amz-date/x-amz-content-sha256 header for AWS REST API.
Example(set to req.*)
        ::

                import awsrest;
                
                backend default {
                  .host = "example-bucket.s3.amazonaws.com";
                }
                
                sub vcl_recv{
                  set req.http.host = "example-bucket.s3.amazonaws.com";
                  awsrest.v4_generic(
                    service           = "s3",
                    region            = "ap-northeast-1",
                    access_key        = "[Your Access Key]",
                    secret_key        = "[Your Secret Key]",
                    signed_headers    = "host;",
                    canonical_headers = "host:" + req.http.host + awsrest.lf()
                  );
                }
                
                //data

                //2 ReqHeader      c Authorization: AWS4-HMAC-SHA256 Credential=****************/20150704/ap-northeast-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=****************
                //2 ReqHeader      c x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
                //2 ReqHeader      c x-amz-date: 20150704T103402Z
                
Example(set to bereq.*)
        ::

                import awsrest;
                
                backend default {
                  .host = "example-bucket.s3.amazonaws.com";
                }
                
                sub vcl_backend_fetch{
                  set bereq.http.host = "example-bucket.s3.amazonaws.com";
                  awsrest.v4_generic(
                    service           = "s3",
                    region            = "ap-northeast-1",
                    access_key        = "[Your Access Key]",
                    secret_key        = "[Your Secret Key]",
                    signed_headers    = "host;",
                    canonical_headers = "host:" + bereq.http.host + awsrest.lf()
                  );
                }
                //data
                //25 BereqHeader    b Authorization: AWS4-HMAC-SHA256 Credential=****************/20150704/ap-northeast-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=****************
                //25 BereqHeader    b x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
                //25 BereqHeader    b x-amz-date: 20150704T103159Z

Example(using session token)
        ::

                import awsrest;
                
                backend default {
                  .host = "example-bucket.s3.amazonaws.com";
                }
                
                sub vcl_backend_fetch{
                  set bereq.http.host = "example-bucket.s3.amazonaws.com";
                  awsrest.v4_generic(
                    service           = "s3",
                    region            = "ap-northeast-1",
                    access_key        = "[Your Access Key]",
                    secret_key        = "[Your Secret Key]",
                    token             = "[Your Session token]",
                    signed_headers    = "host;",
                    canonical_headers = "host:" + bereq.http.host + awsrest.lf()
                  );
                }
                //data
                //25 BereqHeader    b Authorization: AWS4-HMAC-SHA256 Credential=****************/20150704/ap-northeast-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=****************
                //25 BereqHeader    b x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
                //25 BereqHeader    b x-amz-date: 20150704T103159Z
                //25 BereqHeader    b x-amz-security-token: [Your Session Token]


lf
------------------

Prototype
        ::

                lf()
Return value
	STRING
Description
	return LF
Example
        ::

                "x-amz-hoge1:hoge" + awsrest.lf() + "x-amz-hoge2:hoge" + awsrest.lf()


                //data
                x-amz-hoge1:hoge
                x-amz-hoge2:hoge

$Function STRING formurl(STRING url)

Prototype
        ::

                formurl(url)
Return value
	STRING
Description
	Add "=" if field is not have value and delimiter.
	Strip "?","&" from the end of a string.
	AWS signature v4's query-string require sorted field(std.querysort) and field with delimiter(this function)
Example
        ::

                import awsrest;
                import std;
                
                sub vcl_recv{
                  set req.url = awsrest.formurl(std.querysort(req.url));
                }

                //log
                **** v1    0.5 vsl|       1001 ReqURL          c /?aa&cc&bb
                **** v1    0.5 vsl|       1001 ReqURL          c /?aa=&bb=&cc=

INSTALLATION
============

The source tree is based on autotools to configure the building, and
does also have the necessary bits in place to do functional unit tests
using the ``varnishtest`` tool.

Building requires the Varnish header files and uses pkg-config to find
the necessary paths.

Usage::

 ./autogen.sh
 ./configure

If you have installed Varnish to a non-standard directory, call
``autogen.sh`` and ``configure`` with ``PKG_CONFIG_PATH`` pointing to
the appropriate path. For instance, when varnishd configure was called
with ``--prefix=$PREFIX``, use

::

 export PKG_CONFIG_PATH=${PREFIX}/lib/pkgconfig
 export ACLOCAL_PATH=${PREFIX}/share/aclocal

The module will inherit its prefix from Varnish, unless you specify a
different ``--prefix`` when running the ``configure`` script for this
module.

Make targets:

* make - builds the vmod.
* make install - installs your vmod.
* make check - runs the unit tests in ``src/tests/*.vtc``.
* make distcheck - run check and prepare a tarball of the vmod.

If you build a dist tarball, you don't need any of the autotools or
pkg-config. You can build the module simply by running::

 ./configure
 make

Installation directories
------------------------

By default, the vmod ``configure`` script installs the built vmod in the
directory relevant to the prefix. The vmod installation directory can be
overridden by passing the ``vmoddir`` variable to ``make install``.


Google Cloud Storage(GCS) sample
=================================

It can also be used in GCS.
        ::

                import awsrest;
                
                backend default {
                  .host = "example-bucket.storage.googleapis.com";
                }
                
                sub vcl_recv{
                  set req.http.host = "example-bucket.storage.googleapis.com";
                  awsrest.v4_generic(
                    service           = "storage",
                    region            = "asia-northeast1",
                    access_key        = "[Your Access Key]",
                    secret_key        = "[Your Secret Key]",
                    signed_headers    = "host;",
                    canonical_headers = "host:" + req.http.host + awsrest.lf()
                  );
                }


COMMON PROBLEMS
===============

* configure: error: Need varnish.m4 -- see README.rst

  Check if ``PKG_CONFIG_PATH`` has been set correctly before calling
  ``autogen.sh`` and ``configure``

* If you catch signature error in several request(URI-encoded)

  Please check that URI encoded.
  AWS signature v4 is require URI-encode. (ref: http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html#d0e8062 )
  This VMOD does not automatically update update be/req.url.
  Because, can't detect URI-encoded or not.
  
  Sample(replace @ -> %40)::
  
   //////////////////////////
   //In cl-thread.

   sub vcl_recv{
     set req.url = regsuball(req.url,"@","%40");
     awsrest.v4_generic(
       service           = "s3",
       region            = "ap-northeast-1",
       access_key        = "[Your Access Key]",
       secret_key        = "[Your Secret Key]",
       signed_headers    = "host;",
       canonical_headers = "host:" + req.http.host + awsrest.lf()
     );
   }
   //////////////////////////
   //In bg-thread.

   sub vcl_backend_fetch {
     set bereq.url = regsuball(bereq.url,"@","%40");
     awsrest.v4_generic(
       service           = "s3",
       region            = "ap-northeast-1",
       access_key        = "[Your Access Key]",
       secret_key        = "[Your Secret Key]",
       signed_headers    = "host;",
       canonical_headers = "host:" + bereq.http.host + awsrest.lf()
     );
   }

* If a signature error occurs when using a query-string

  AWS signature v4's query-string require sorted field and field with delimiter.
  
  Failed url::
  
   /a?c=1&b=1
   /a?b&c
 
  Success url::

    /a?b=1&c=1
    /a?b=&c=
  
  Use std.querysort and awsrest.formurl to solve it.
  
  Sample::
  
   sub vcl_recv{
     set req.url = awsrest.formurl(std.querysort(req.url));
     awsrest.v4_generic(
       service           = "s3",
       region            = "ap-northeast-1",
       access_key        = "[Your Access Key]",
       secret_key        = "[Your Secret Key]",
       signed_headers    = "host;",
       canonical_headers = "host:" + req.http.host + awsrest.lf()
     );
   }
 

COPYRIGHT
=============

This document is licensed under the same license as the
libvmod-awsrest project. See LICENSE for details.

* Copyright (c) 2012-2021 Shohei Tanaka(@xcir)

File layout and configuration based on libvmod-example

* Copyright (c) 2011 Varnish Software AS
* https://github.com/varnishcache/libvmod-example/
* https://github.com/varnishcache/libvmod-example/blob/master/LICENSE

hmac-sha1 and base64 based on libvmod-digest

* Copyright (c) 2011-2019 Varnish Software AS
* https://github.com/varnish/libvmod-digest
* https://github.com/varnish/libvmod-digest/blob/master/src/vmod_digest.c

headersort based on libvmod-std/querysort

* Copyright (c) 2010-2014 Varnish Software AS
* https://github.com/varnishcache/varnish-cache
* https://github.com/varnishcache/varnish-cache/blob/master/vmod/vmod_std_querysort.c


CHANGES.rst
This is a running log of changes to libvmod-awsrest.

libvmod-awsrest 70.12 (2021-10-19)
--------------------------------

* Fixed sorting problems in Signed/CanonicalHeaders (#32)

libvmod-awsrest 70.11 (2021-10-18)
--------------------------------

* Support Varnish 6.6.x and 7.0.x

libvmod-awsrest 62.10 (2019-03-20)
--------------------------------

* Support Varnish 6.2.x
* Change ABI check from strict to vrt.

libvmod-awsrest 60.9 (2018-06-22)
--------------------------------

* Support Varnish 6.0.x

libvmod-awsrest 51.8 (2017-07-27)
--------------------------------

* Add .formurl function.
  (Thanks issue #24 Kaelten)

libvmod-awsrest 51.7 (2017-04-20)
--------------------------------

* Drop support Varnish 4.1.x
  Sync newst vmod_example

libvmod-awsrest 50.6 (2017-04-19)
--------------------------------

* Support session token.
  IMPORTANT: Added token parameter at v4_generic.
  (Thanks p-r #20 dgoodlad)

libvmod-awsrest 50.5 (2016-11-19)
--------------------------------

* Support Varnish 4.1.x / 5.0.x
  (Thanks p-r #12 poblahblahblah, issue #11 huayra)

libvmod-awsrest 0.4 (2015-07-04)
--------------------------------

* Support Varnish 4.0.x

libvmod-awsrest 0.3 (2015-05-06)
--------------------------------

* Support AWS V4 Signature
* Delete method for v1 signature.

libvmod-awsrest 0.2 (2013-01-14)
--------------------------------

* Added s3_generic_iam
  (Thanks p-r #1 RevaxZnarf)

libvmod-awsrest 0.1 (2012-07-04)
--------------------------------

First release




src/tests/vta.php
<?php

//
//php vta.php [aws config] [region] [bucket] [url]
//php vta.php ~/.aws/config ap-northeast-1 examplebucket /foo/bar


//non-sts
$r   = parse_ini_file($argv[1]);
$cmd = sprintf('varnishtest -Dregion=%s -Daccesskey="%s" -Dsecretkey="%s" -Dbucket="%s" -Durl="%s" r*.vta',
         $argv[2],
         $r['aws_access_key_id'],
         $r['aws_secret_access_key'],
         $argv[3],
         $argv[4]
  );
echo shell_exec($cmd);

//sts
$r   = json_decode(shell_exec('aws sts get-session-token'), 1);
$i   = ceil(strlen($r['Credentials']['SessionToken'])/2);
$cmd = sprintf('varnishtest -Dregion=%s -Daccesskey="%s" -Dsecretkey="%s" -Dtoken1="%s" -Dtoken2="%s" -Dbucket="%s" -Durl="%s" t*.vta',
         $argv[2],
         $r['Credentials']['AccessKeyId'],
         $r['Credentials']['SecretAccessKey'],
         substr($r['Credentials']['SessionToken'],0,$i),
         substr($r['Credentials']['SessionToken'],$i),
         $argv[3],
         $argv[4]
  );
echo shell_exec($cmd);




src/tests/vtg.php
<?php

//
//php vta.php [region] [bucket] [url]
//php vtg.php ~/.aws/config asia-northeast1 examplebucket /foo/bar


//non-sts
$r   = parse_ini_file($argv[1]);
$cmd = sprintf('varnishtest -Dregion=%s -Daccesskey="%s" -Dsecretkey="%s" -Dbucket="%s" -Durl="%s" r*.vtg',
         $argv[2],
         $r['aws_access_key_id'],
         $r['aws_secret_access_key'],
         $argv[3],
         $argv[4]
  );
echo shell_exec($cmd);




src/tests/vtx.php
<?php

//
//php vtx.php [aws config] [region] [bucket] [url]
//php vtx.php ~/.aws/config ap-northeast-1 examplebucket /foo/bar


//non-sts
$r   = parse_ini_file($argv[1]);
$cmd = sprintf('varnishtest -Dregion=%s -Daccesskey="%s" -Dsecretkey="%s" -Dbucket="%s" -Durl="%s" r*.vtx',
         $argv[2],
         $r['aws_access_key_id'],
         $r['aws_secret_access_key'],
         $argv[3],
         $argv[4]
  );
echo shell_exec($cmd);

//sts
$r   = json_decode(shell_exec('aws sts get-session-token'), 1);
$i   = ceil(strlen($r['Credentials']['SessionToken'])/2);
$cmd = sprintf('varnishtest -Dregion=%s -Daccesskey="%s" -Dsecretkey="%s" -Dtoken1="%s" -Dtoken2="%s" -Dbucket="%s" -Durl="%s" t*.vtx',
         $argv[2],
         $r['Credentials']['AccessKeyId'],
         $r['Credentials']['SecretAccessKey'],
         substr($r['Credentials']['SessionToken'],0,$i),
         substr($r['Credentials']['SessionToken'],$i),
         $argv[3],
         $argv[4]
  );
echo shell_exec($cmd);




src/vmod_awsrest.c
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/* need vcl.h before vrt.h for vmod_evet_f typedef */
#include "cache/cache.h"
#include "vcl.h"

#ifndef VRT_H_INCLUDED
#include <vrt.h>
#endif

#ifndef VDEF_H_INCLUDED
#include <vdef.h>
#endif

#include "vtim.h"
#include "vcc_awsrest_if.h"

#include <time.h>
#include <string.h>
#include <stdio.h>
#include <mhash.h>

int
#if VRT_MAJOR_VERSION > 8U
  vmod_event_function(VRT_CTX, struct vmod_priv *priv, enum vcl_event_e e)
#else
  event_function(VRT_CTX, struct vmod_priv *priv, enum vcl_event_e e)
#endif
{
	return (0);
}

static int
compa(const void *a, const void *b)
{
	const char * const *pa = a;
	const char * const *pb = b;
	const char *a1, *b1;

	for (a1 = pa[0], b1 = pb[0]; a1 < pa[1] && b1 < pb[1]; a1++, b1++)
		if (*a1 != *b1)
			return (*a1 - *b1);
	return (0);
}

char * 
headersort(VRT_CTX, char *txt, char sep, char sfx)
{
	const char *cq, *cu;
	char *p, *r;
	const char **pp;
	const char **pe;
	unsigned u;
	int np;
	int i;

	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);

	if (txt == NULL)
		return (NULL);

	/* Split :query from :url */
	cu = txt;

	/* Spot single-param queries */
	cq = strchr(cu, sep);
	if (cq == NULL)
		return (txt);

	r = WS_Copy(ctx->ws, txt, -1);
	if (r == NULL)
		return (txt);

	u = WS_ReserveLumps(ctx->ws, sizeof(const char **));
#if VRT_MAJOR_VERSION >= 12U
	pp = WS_Reservation(ctx->ws);
#else
	pp = (const char**)(void*)(ctx->ws->f);
#endif
	if (u < 4) {
		WS_Release(ctx->ws, 0);
		WS_MarkOverflow(ctx->ws);
		return (txt);
	}
	pe = pp + u;

	/* Collect params as pointer pairs */
	np = 0;
	pp[np++] =  cu;
	for (cq =  cu; *cq != '\0'; cq++) {
		if (*cq == sep) {
			if (pp + np + 3 > pe) {
				WS_Release(ctx->ws, 0);
				WS_MarkOverflow(ctx->ws);
				return (txt);
			}
			pp[np++] = cq;
			/* Skip trivially empty params */
			while (cq[1] == sep)
				cq++;
			pp[np++] = cq + 1;
		}
	}
	pp[np++] = cq;
	assert(!(np & 1));

	qsort(pp, np / 2, sizeof(*pp) * 2, compa);

	/* Emit sorted params */
	p =  r + (cu - txt);
	cq = "";
	for (i = 0; i < np; i += 2) {
		/* Ignore any edge-case zero length params */
		if (pp[i + 1] == pp[i])
			continue;
		assert(pp[i + 1] > pp[i]);
		if (*cq)
			*p++ = *cq;
		memcpy(p, pp[i], pp[i + 1] - pp[i]);
		p += pp[i + 1] - pp[i];
		cq = &sep;
	}
	if(sfx){
		*p = sfx;
		p++;
	}
	*p = '\0';

	WS_Release(ctx->ws, 0);
	return (r);
}






/////////////////////////////////////////////
static const char *
vmod_hmac_sha256(VRT_CTX,
	const char *key,size_t lkey, const char *msg,size_t lmsg,bool raw)
{
	hashid hash = MHASH_SHA256;
	size_t blocksize = mhash_get_block_size(hash);

	char *p;
	char *ptmp;
	p    = WS_Alloc(ctx->ws, blocksize * 2 + 1);
	ptmp = p;

	
	unsigned char *mac;
	unsigned u;
	#if VRT_MAJOR_VERSION > 9
	u = WS_ReserveAll(ctx->ws);
	#else
	u = WS_Reserve(ctx->ws, 0);
	#endif
	assert(u > blocksize);
	mac = (unsigned char*)ctx->ws->f;
	
	int i;
	MHASH td;

	assert(msg);
	assert(key);

	assert(mhash_get_hash_pblock(hash) > 0);

	td = mhash_hmac_init(hash, (void *) key, lkey,
		mhash_get_hash_pblock(hash));
	mhash(td, msg, lmsg);
	mhash_hmac_deinit(td,mac);
	if(raw){
		WS_Release(ctx->ws, blocksize);
		return (char *)mac;
	}
	WS_Release(ctx->ws, 0);
	
	for (i = 0; i<blocksize;i++) {
		sprintf(ptmp,"%.2x",mac[i]);
		ptmp+=2;
	}
	return p;
}

static const char *
vmod_v4_getSignature(VRT_CTX,
	const char* secret_key, const char* dateStamp, const char* regionName, const char* serviceName,const char* string_to_sign
){
	size_t len = strlen(secret_key) + 5;
	char key[len];
	char *kp = key;
	sprintf(kp,"AWS4%s",secret_key);
	
	const char *kDate    = vmod_hmac_sha256(ctx,kp,strlen(kp), dateStamp,strlen(dateStamp),true);
	const char *kRegion  = vmod_hmac_sha256(ctx,kDate,   32, regionName,strlen(regionName),true);
	const char *kService = vmod_hmac_sha256(ctx,kRegion, 32, serviceName,strlen(serviceName),true);
	const char *kSigning = vmod_hmac_sha256(ctx,kService,32, "aws4_request", 12,true);
	
	return vmod_hmac_sha256(ctx,kSigning,32, string_to_sign,strlen(string_to_sign),false);
}


static const char *
vmod_hash_sha256(VRT_CTX, const char *msg)
{
	MHASH td;
	hashid hash = MHASH_SHA256;
	unsigned char h[mhash_get_block_size(hash)];
	int i;
	char *p;
	char *ptmp;
	td = mhash_init(hash);
	mhash(td, msg, strlen(msg));
	mhash_deinit(td, h);
	p = WS_Alloc(ctx->ws,mhash_get_block_size(hash)*2 + 1);
	ptmp = p;
	for (i = 0; i<mhash_get_block_size(hash);i++) {
		sprintf(ptmp,"%.2x",h[i]);
		ptmp+=2;
	}
	return p;
}
void vmod_v4_generic(VRT_CTX,
	VCL_STRING service,               //= 's3';
	VCL_STRING region,                //= 'ap-northeast-1';
	VCL_STRING access_key,            //= 'your access key';
	VCL_STRING secret_key,            //= 'your secret key';
	VCL_STRING token,                 //= 'optional session token';
	VCL_STRING signed_headers,       //= 'host;';// x-amz-content-sha256;x-amz-date is appended by default.
	VCL_STRING canonical_headers,    //= 'host:s3-ap-northeast-1.amazonaws.com\n'
	VCL_BOOL feature                  //= reserved param(for varnish4)
){
	////////////////
	//get data
	const char *method;
	const char *requrl;
	struct http *hp;
	struct gethdr_s gs;
	
	if (ctx->http_bereq !=NULL && ctx->http_bereq->magic== HTTP_MAGIC){
		//bg-thread
		hp = ctx->http_bereq;
		gs.where = HDR_BEREQ;
	}else{
		//cl-thread
		hp = ctx->http_req;
		gs.where = HDR_REQ;
	}
	method= hp->hd[HTTP_HDR_METHOD].b;
	requrl= hp->hd[HTTP_HDR_URL].b;

	////////////////
	//create date
	time_t tt;
	char amzdate[33];
	char datestamp[25];
	tt = time(NULL);
	struct tm * gmtm = gmtime(&tt);
	
	sprintf(amzdate,
		"%d%02d%02dT%02d%02d%02dZ",
		gmtm->tm_year +1900,
		gmtm->tm_mon  +1,
		gmtm->tm_mday,
		gmtm->tm_hour,
		gmtm->tm_min,
		gmtm->tm_sec
	);
	sprintf(datestamp,
		"%d%02d%02d",
		gmtm->tm_year +1900,
		gmtm->tm_mon  +1,
		gmtm->tm_mday
	);

	////////////////
	//create payload
	const char * payload_hash = vmod_hash_sha256(ctx, "");
	
	////////////////
	//create signed headers
	size_t tokenlen = 0;
	if(token != NULL) tokenlen = strlen(token);

	size_t len = strlen(signed_headers) + 32;
	if(tokenlen > 0) len += 21; // ;x-amz-security-token
	char *psh = WS_Alloc(ctx->ws,len);
	char *psigned_headers = WS_Alloc(ctx->ws,len);
	if(tokenlen > 0) {
		sprintf(psh,"%sx-amz-content-sha256;x-amz-date;x-amz-security-token",signed_headers);
	} else {
		sprintf(psh,"%sx-amz-content-sha256;x-amz-date",signed_headers);
	}
	psigned_headers = headersort(ctx, psh, ';', 0);
	////////////////
	//create canonical headers
	len = strlen(canonical_headers) + 115;
	// Account for addition of "x-amz-security-token:[token]\n"
	if(tokenlen > 0) len += 22 + tokenlen;
	char *pch = WS_Alloc(ctx->ws,len);
	char *pcanonical_headers = WS_Alloc(ctx->ws,len);

	if(tokenlen > 0) {
		sprintf(pch,"%sx-amz-content-sha256:%s\nx-amz-date:%s\nx-amz-security-token:%s\n",canonical_headers,payload_hash,amzdate,token);
	} else {
		sprintf(pch,"%sx-amz-content-sha256:%s\nx-amz-date:%s\n",canonical_headers,payload_hash,amzdate);
	}
	pcanonical_headers = headersort(ctx, pch, '\n', '\n');

	////////////////
	//create credential scope
	len = strlen(datestamp)+ strlen(region)+ strlen(service)+ 16;
	char *pcredential_scope = WS_Alloc(ctx->ws,len);
	sprintf(pcredential_scope,"%s/%s/%s/aws4_request",datestamp,region,service);
	
	////////////////
	//create canonical request
	len = strlen(method)+ strlen(requrl)+ strlen(pcanonical_headers)+ strlen(psigned_headers)+ strlen(payload_hash) + 6;
	char *pcanonical_request = WS_Alloc(ctx->ws,len);
	char tmpform[32];
	tmpform[0]=0;
	char *ptmpform = &tmpform[0];

	char *adr = strchr(requrl, (int)'?');
	if(adr == NULL){
		sprintf(pcanonical_request,"%s\n%s\n\n%s\n%s\n%s",
			method,
			requrl,
			pcanonical_headers,
			psigned_headers,
			payload_hash
		);
	}else{
		sprintf(ptmpform,"%s.%lds\n%s","%s\n%",(adr - requrl),"%s\n%s\n%s\n%s");
		sprintf(pcanonical_request,ptmpform,
			method,
			requrl,
			adr + 1,
			pcanonical_headers,
			psigned_headers,
			payload_hash
		);
	}
	
	
	////////////////
	//create string_to_sign
	len = strlen(amzdate)+ strlen(pcredential_scope)+ 33;
	char *pstring_to_sign = WS_Alloc(ctx->ws,len);
	sprintf(pstring_to_sign,"AWS4-HMAC-SHA256\n%s\n%s\n%s",amzdate,pcredential_scope,vmod_hash_sha256(ctx, pcanonical_request));
	
	////////////////
	//create signature
	const char *signature = vmod_v4_getSignature(ctx,secret_key,datestamp,region,service,pstring_to_sign);

	////////////////
	//create authorization
	len = strlen(access_key)+ strlen(pcredential_scope)+ strlen(psigned_headers)+ strlen(signature)+ 58;
	char *pauthorization= WS_Alloc(ctx->ws,len);
	
	sprintf(pauthorization,"AWS4-HMAC-SHA256 Credential=%s/%s, SignedHeaders=%s, Signature=%s",
		access_key,
		pcredential_scope,
		psigned_headers,
		signature);
	
#if VRT_MAJOR_VERSION >= 14U
        ////////////////
        //Set to header
        gs.what = "\016Authorization:";
        VRT_SetHdr(ctx, &gs        , pauthorization , NULL);
        gs.what = "\025x-amz-content-sha256:";
        VRT_SetHdr(ctx, &gs , payload_hash , NULL);
        gs.what = "\013x-amz-date:";
        VRT_SetHdr(ctx, &gs           , amzdate , NULL);
        if(tokenlen > 0){
          gs.what="\025x-amz-security-token:";
          VRT_SetHdr(ctx, &gs, token, NULL);
        }
#else
        ////////////////
        //Set to header
        gs.what = "\016Authorization:";
        VRT_SetHdr(ctx, &gs        , pauthorization , vrt_magic_string_end);
        gs.what = "\025x-amz-content-sha256:";
        VRT_SetHdr(ctx, &gs , payload_hash , vrt_magic_string_end);
        gs.what = "\013x-amz-date:";
        VRT_SetHdr(ctx, &gs           , amzdate , vrt_magic_string_end);
        if(tokenlen > 0){
          gs.what="\025x-amz-security-token:";
          VRT_SetHdr(ctx, &gs, token, vrt_magic_string_end);
        }
#endif
}

VCL_STRING
vmod_lf(VRT_CTX){
	char *p;
	p = WS_Alloc(ctx->ws,2);
	strcpy(p,"\n");
	return p;
}


VCL_STRING
vmod_formurl(VRT_CTX, VCL_STRING url){
	char *adr, *ampadr, *eqadr;
	char *pp, *p;
	unsigned u;
	int len = 0;
	int cnt = 0;
	const char *lst= url + strlen(url) -1;

	adr = strchr(url, (int)'?');
	
	if(adr == NULL){
		return url;
	}

	#if VRT_MAJOR_VERSION > 9
	u = WS_ReserveAll(ctx->ws);
	#else
	u = WS_Reserve(ctx->ws, 0);
	#endif
	pp = p = ctx->ws->f;

	len = adr - url;
	if(len + 1 > u){ // 1=(null)
		WS_Release(ctx->ws, 0);
		WS_MarkOverflow(ctx->ws);
		return url;
	}
	memcpy(p, url, len);
	p+=len;
	for(;lst >url;lst--){
		if(*lst != '?' && *lst != '&'){
			lst++;
			break;
		}
	}
	if(lst <= adr){
		// url: /xxxx? /?
		*p = 0;
		p++;
		WS_Release(ctx->ws, p - pp);
		return(pp);
	}
	
	while(1){
		ampadr = memchr(adr +1, (int)'&', lst - adr -1);
		if(ampadr == NULL){
			len = lst - adr;
			if(p - pp + len + 2 > u){ // 2= strlen("=")+1(null)
				WS_Release(ctx->ws, 0);
				WS_MarkOverflow(ctx->ws);
				return url;
			}
			memcpy(p, adr, len);
			p+=len;
			
			eqadr = memchr(adr +1, (int)'=', lst - adr -1);
			if(eqadr == NULL){
				cnt++;
				*p = '=';
				p++;
			}
			break;
		}else{
			eqadr = memchr(adr +1, (int)'=', ampadr - adr -1);
			len = ampadr - adr;
			if(p - pp + len + 2 > u){
				WS_Release(ctx->ws, 0);
				WS_MarkOverflow(ctx->ws);
				return url;
			}
			memcpy(p, adr, len);
			p+=len;
			if(eqadr == NULL){
				cnt++;
				*p = '=';
				p++;
			}
			adr = ampadr;
		}
	}
	*p = 0;
	p++;
	WS_Release(ctx->ws, p - pp);
	
	return(pp);
	
}
